MAINT: Implemented two dtype-related TODO's

BvB93 opened the pull request:
This pull requests addresses and implement two previous TODO's in `np.typing._DTypeLike`:
* Added a protocol, `_SupportsDtype`, for objects with the `dtype` attribute.
* Replaced an old dictionary, representing array field parameter, with a `TypedDict`.

Commit:
Added a `TypedDict` to the `_DtypeLike` union

Commit:
Updated dtype-related tests

Commit:
Ran black

Commit:
Passing (type, 1) or '1type' is deprecated; remove it from `_DtypeLike`

Commit:
Added the `_SupportsDtype` protocol

Commit:
Partially reverted 599e02dcdb12c96eeeb258c31576e0a5ca0bc8a2

Commit:
Removed a now-redundant comment

eric-wieser commented on numpy/typing.pyi:
```diff
@@ -15,6 +15,22 @@ _ShapeLike = Union[int, Sequence[int]]
+# Mandatory + optional keys
+class _DtypeDict(_DtypeDictBase, total=False):
+    offsets: Sequence[int]
+    titles: Sequence[Union[bytes, Text, None]]
```
```suggestion
    titles: Sequence[Any]
```
`np.dtype(dict(names=['a'], formats=[int], titles=[object()]))` is apparently allowed.

BvB93 replied:
Good call; fixed in https://github.com/numpy/numpy/pull/16622/commits/a96f1f1ab740c2509720df98c535f0034ed453de.

Commit:
Implement https://github.com/numpy/numpy/pull/16622#discussion_r441346506

The `"titles"` key in `_DtypeDict' can take a sequence of any hashable object as argument.

eric-wieser commented on numpy/typing.pyi:
```diff
@@ -23,7 +23,7 @@ class _DtypeDictBase(TypedDict):
 class _DtypeDict(_DtypeDictBase, total=False):
     offsets: Sequence[int]
-    titles: Sequence[Union[bytes, Text, None]]
+    titles: Sequence[Any]  # Elements should be valid as dict keys
```
This comment isn't accurate. _If_ the elements are `str`, then they end up in `dtype.fields`. If they're not, then they don't.

There are no requirements on hashability, nor will numpy even attempt to hash them if they are not `str`.

Perhaps
```suggestion
    titles: Sequence[Any]  # Only `str` elements are usable as indexing aliases, but all objects are legal
```

BvB93 replied:
Ah, you're right. After reading the [structured arrays](https://numpy.org/doc/stable/user/basics.rec.html#field-titles) documentation I was under the impression that passed values would always serve as dict `dtype.fields` keys.

Still, this does beg the question whether or not there is actually any legit use for non-`str` keys here. Do they actually serve a purpose or are they simply ignore?
If it's the latter case, shouldn't we actively discourage their use?

BvB93 replied:
Any thoughts on the concern I raised above?

BvB93 replied:
Updated the respective comment (https://github.com/numpy/numpy/pull/16622/commits/b4fa0b966732124afce2d19b460ee35a5b2ce124).

Commit:
Update _dtype_like.py

Commit:
Fix two issues which arose during the latest rebase

* The `Protocol` baseclass of `_SupportsDtype` disappeared.
* Ensure that `typing_extensions` is not imported during runtime, as it's not a hard NumPy dependency.

Commit:
Updated tests

BvB93 commented:
During the latest rebase two issues were introduced by accident.
They have been resolved now (see https://github.com/numpy/numpy/pull/16622/commits/9d16ab7a3b98d8dabf62ec2e0e7cca7d190fd795 for more details).

anirudh2290 commented on numpy/typing/_dtype_like.py:
```diff
@@ -28,17 +56,7 @@
-    ],
+    _DtypeDict,
     # {'field1': ..., 'field2': ..., ...}
     Dict[str, Tuple[_DtypeLikeNested, int]],
```
maybe not related to this PR, but looks like this type of usage is discouraged in numpy ( https://numpy.org/devdocs/reference/arrays.dtypes.html ). Would static types be a good place to error on this usage if it is not encouraged ?

BvB93 replied:
> Would static types be a good place to error on this usage if it is not encouraged ?

I would argue that it is, though I feel such a change will require some discussion beforehand.

anirudh2290 replied:
okay, i will open an issue.

anirudh2290 replied:
added an issue: https://github.com/numpy/numpy/issues/16891

anirudh2290 replied:
Based on the discussion in #16891, we can remove this line and move https://github.com/numpy/numpy/pull/16622/files#diff-ccb6649c3fc407a2b5413ef6f80f1dfcR28 to numpy/tests/typing/fail/dtype.py ?

BvB93 replied:
And done as of https://github.com/numpy/numpy/pull/16622/commits/1c701a7cbc51c04d30d35767adc78863fabe4542.

anirudh2290 commented on numpy/tests/typing/pass/dtype.py:
```diff
@@ -1,3 +1,33 @@
+    dtype = float
+
+
+np.dtype(Test())
```
maybe worth adding a list example here like : np.dtype([('name', np.unicode_, 16), ('grades', np.float64, (2,))]).

Didnt find it in the tests.

BvB93 replied:
Good call! Implemented in https://github.com/numpy/numpy/pull/16622/commits/12333c72aee600ba66387f7963dd73b4801e0fc0.

anirudh2290 reviewed (approved):
Changes specific to this PR LGTM !

Commit:
Added a test for `dtype([...])` input

See https://github.com/numpy/numpy/pull/16622#discussion_r455440244

Commit:
MAINT: Removed support for the `np.dtype({'field1': ..., 'field2': ..., ...})` syntax

Removed per the discussion in https://github.com/numpy/numpy/issues/16891

person142 commented on numpy/typing/_dtype_like.py:
```diff
@@ -1,10 +1,37 @@
+        dtype: _DtypeLikeNested
+
+else:  # runtime-only placeholders
+    _DtypeDict = 'numpy.typing._dtype_like._DtypeDict'
```
We had previously decided against using runtime-only placeholders unless it was 100% necessary; i.e. `sys.version_info < (3, 8, )` and `typing_extensions` not installed. Why break from that here?

BvB93 replied:
> Why break from that here?

An oversight I'm afraid. Fixed in https://github.com/numpy/numpy/pull/16622/commits/fd796670d9b9cc14dc6a98a70cfe495e72e5691e.

Commit:
MAINT: don't use runtime-only placeholders unless necessary

Addresses https://github.com/numpy/numpy/pull/16622#discussion_r462719703

BvB93 commented:
Rebased to get rid of the circleci failure.

Commit:
MAINT: clarified a comment

Addresses https://github.com/numpy/numpy/pull/16622#discussion_r441455084

Co-Authored-By: Eric Wieser <eric-wieser@users.noreply.github.com>

BvB93 commented:
If there are no more comments, remarks, _etc._ then let's get this merged soon, as https://github.com/numpy/numpy/pull/16917 is also waiting on this pull request.

person142 commented:
I’ll take another look tonight.

person142 commented:
Looks good to me, and test failure is unrelated so merging. Thanks @BvB93.
